docs/adr/0007-seperate-binding-generators.md

```diff
@@ -0,0 +1,136 @@
+- `uniffi_bindgen` staying as the source of truth (with it passing it's version to the other crates) has a good benefit of reducing metal load to keep track of which binary to call
+
+
+### **[Option 3] Publish a separate crate per binding generator and have them get consumed directly**
+
+* Good, because maintainability will be clear, and members of the community can opt to maintain their own binding generators
+* Good, because testability can improve, where a developer can test a binding generator without building or testing the others
+* Good, because our CI would only need to test the core bindings we maintain, and others can be tested by their own maintainers (for example, a `gecko-js` binding generator should be tested in `mozilla-central` and not here)
+* Good, because a release in one binding wouldn't have an impact on any other ones unless it changes internal `uniffi` behavior.
+* Bad, because it's easier to accidentally have a version mismatch. And we have even less flexibility than `Option 2` to mitigate it since `uniffi_bindgen` won't shell out to the crate.
+* Bad, because it's not clear how the users should generate the scaffolding, and if they use `uniffi_bindgen` there is a risk of version incompatibility.
+* Bad, because testability might increase in complexity. Bindings not in this repository will not have access to the extensive fixtures and examples we have. We might need to expose those if take this option.
+* Bad because it increases the number of crates we manage. As we would manage the core crates for the bindings we depend on (i.e kotlin, swift, python, and soon a C++/gecko one). This will make the release process painful, but we can add automation and clear crate ownership to help.
+
+
+## Decision Outcome
+
+Chosen option:
+
+* **[Option 2] Publish a separate crate per binding generator and have users use the  `uniffi_bindgen` Command line interface to shell out to the specific binding generators**
+
+### Changes Proposed
+#### The exposed trait for binding generators to implement
+`unffi_bindgen` would define a public trait, `BindingGenerator` that includes one main function:
+```rs
+/// Takes in the ComponentInterface and Config and writes the bindings into directory `out_dir`
+fn write_bindings<P: AsRef<Path>>(&self, ci: &ComponentInterface, config: Config, out_dir: P) -> Result<()>;
+```
+
+Additionally, we can define optional functions to extend testing functionality
+```rs
+/// Compiles the generated bindings using `write_bindings`
+fn compile_bindings<P: AsRef<Path>>(&self, ci: &ComponentInterface, out_dir: P) -> Result<()>;
+
+/// Runs a script in the language that implements the `BindingGenerator` trait
+fn run_script<P: AsRef<Path>>(&self, out_dir: P, script_file: P) -> Result<()>;
```

Comment:
I don't quite understand this. It seems you are suggesting this method is declared on the `BindingGenerator` trait, but it also runs a script that implements the `BindingGenerator` trait?

(I'm trying to think a little about how this might work for a `gecko-js` binding - `compile_bindings` might generate webidl and/or some header files? `run_script` would be used for tests? Or for a build? Or maybe I'm missing something :)

Reply:
Yeah `run_script` is purely for testing purposes - the comment up there is a little misleading, it should say "Runs a script in the target language, useful for running tests"